src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -305,28 +366,107 @@ export class CodeActionMenu extends Disposable implements IEditorContribution {
 		renderDisposables.add(this._editor.onDidLayoutChange(e => this.hideCodeActionWidget()));
 
-		// Populating the list widget and tracking enabled options.
+		enum CodeActionArrayGroup {
+			quickfix = 0,
+			extract = 1,
+			convert = 2,
+			normalRefactor = 3,
+			separator = 4,
+			other = 5,
+		}
+
+		// Filters and groups code actions by their group
+		const menuEntries: IAction[][] = [[], [], [], [], [], []];
+		let groupID = 0;
 		inputArray.forEach((item, index) => {
+			if (item instanceof CodeActionAction) {
+				// create codea action kind from string
+				// not array, but class -> add into class
+				// TODO:
+				const optionKind = item.action.kind;
+
+				if (CodeActionKind.QuickFix.contains(new CodeActionKind(String(optionKind)))) {
+					groupID = CodeActionArrayGroup.quickfix;
+				} else if (optionKind?.startsWith(CodeActionGroupID.extract)) {
+					groupID = CodeActionArrayGroup.extract;
+				} else if (optionKind?.startsWith(CodeActionGroupID.convert)) {
+					groupID = CodeActionArrayGroup.convert;
+				} else if (optionKind === CodeActionGroupID.normalRefactor) {
+					groupID = CodeActionArrayGroup.normalRefactor;
+				} else if (item.id === `vs.actions.separator`) {
+					groupID = CodeActionArrayGroup.separator;
+				} else {
+					groupID = CodeActionArrayGroup.other;
+				}
 
-			const currIsSeparator = item.class === 'separator';
+			} else if (item.id === `vs.actions.separator`) {
+				groupID = CodeActionArrayGroup.separator;
+			}
 
-			if (currIsSeparator) {
-				// set to true forever because there is a separator
-				this.hasSeparator = true;
+			menuEntries[groupID].push(item);
+		});
+
+		// Creates flat list of all menu entries with headers as separators
+		let numHeaders = 0;
+		const totalActionEntries: (IAction | string)[] = [];
+		menuEntries.forEach(entry => {
+			if (entry.length > 0 && entry[0] instanceof CodeActionAction) {
+				const firstAction = entry[0].action.kind;
+				if (CodeActionKind.QuickFix.contains(new CodeActionKind(String(firstAction)))) {
+					totalActionEntries.push(localize('codeAction.widget.id.quickfix', 'Quickfix'));
+					totalActionEntries.push(...entry);
+					numHeaders++;
+				} else if (firstAction?.startsWith(CodeActionGroupID.extract)) {
+					totalActionEntries.push(localize('codeAction.widget.id.extract', 'Extract'));
+					totalActionEntries.push(...entry);
+					numHeaders++;
+				}
+				else if (firstAction?.startsWith(CodeActionGroupID.convert)) {
+					totalActionEntries.push(localize('codeAction.widget.id.convert', 'Convert'));
+					totalActionEntries.push(...entry);
+					numHeaders++;
+				}
+				else if (firstAction === CodeActionGroupID.normalRefactor) {
+					totalActionEntries.push(localize('codeAction.widget.id.surround', 'Surround With ...'));
+					totalActionEntries.push(...entry);
+					numHeaders++;
+				} else {
+					totalActionEntries.push(...entry);
```

Comment:
If something falls in `other`, do we need to still insert a header so that it isn't accidentally listed under the previous header

Reply:
At the moment, I capture all code aciton cases at the top (Quickfix, Refactor, source action). I add checks for documentation as well, and can remove this for now. Not sure if there are any other menuActions that could potentially not be caught by the various checks i have here